### PR TITLE
vmm: Manually override CPU physical bits

### PR DESCRIPTION
**Pull request by sboeuf**
The goal for this pull request is to provide the user with the ability to limit the guest address space to a specific value.
By introducing new option `phys_bits` to the `--cpus` parameter, one can specify the amount of physical bits exposed to the guest, which will have the effect of limiting the CPU addressable space as well as the available address space for devices.

### CODE REVIEW

File: vmm/src/config.rs [hunk 1]
```diff
@@ -346,12 +349,16 @@ impl CpusConfig {
             .map_err(Error::ParseCpus)?
             .unwrap_or(Toggle(false))
             .0;
+        let phys_bits = parser
+            .convert::<u8>("phys_bits")
+            .map_err(Error::ParseMemoryZone)?;
```
Comment (liuw) [inline vmm/src/config.rs:354]: Shouldn't this be `Error::ParseCpus`?

Comment (sboeuf) [reply vmm/src/config.rs:354]: Good catch! Copy and pasta error :)

File: arch/src/x86_64/mod.rs [hunk 2]
```diff
@@ -410,6 +411,15 @@ pub fn configure_vcpu(
             .map_err(|_| Error::PopulatingCpuid)?;
     }
 
+    // Override CPU physical bits
+    if let Some(phys_bits) = phys_bits {
+        for entry in cpuid.as_mut_slice().iter_mut() {
+            if entry.function == 0x8000_0008 {
+                entry.eax = (entry.eax & 0xffff_ff00) | (phys_bits as u32 & 0xff);
```
Comment (liuw) [inline arch/src/x86_64/mod.rs:418]: I'm wondering if we should somehow check here or earlier the specified value is not larger than the host can support. The provides a small safety net in case people specify too large a value.

(This is not a blocking comment.)

Comment (liuw) [reply arch/src/x86_64/mod.rs:418]: For example, in the CI patch you specify the value to 40. I wonder what happens if I run that locally on a desktop class machine whose physical address size is 39 bits.

Comment (liuw) [reply arch/src/x86_64/mod.rs:418]: I just tried setting the value to 40 on a host with 32 bits physical. Windows still boots. Interesting.

File: vmm/src/config.rs [hunk 3]
```diff
@@ -320,6 +320,8 @@ pub struct CpusConfig {
     pub topology: Option<CpuTopology>,
     #[serde(default)]
     pub kvm_hyperv: bool,
+    #[serde(default)]
+    pub phys_bits: Option<u8>,
```
Comment (sameo) [inline vmm/src/config.rs:324]: This is the user's *desired* physical bits, or `max_phys_bits` because it's the maximum number of bits that the caller is ready to use with its workload. I think the logic should be that the actual physical bits that we're going to use is:

If `max_phys_bits` is `None` then `actual_phys_bits = get_host_cpu_phys_bit()` else `actual_physical_bits = min(config.max_phys_bits, get_host_cpu_phys_bits())`.

Then `actual_phys_bits` would be the number that you'd use for the memory allocator and the CPUID patch.

File: arch/src/x86_64/mod.rs (hunk 2 above)
Comment (sameo) [reply arch/src/x86_64/mod.rs:418]: I think this one is for correctness purposes, there's nothing making it mandatory for the guest to read this leaf.

File: vmm/src/config.rs (hunk 3 above)
Comment (sboeuf) [reply vmm/src/config.rs]: Done!

File: src/main.rs [hunk 4]
```diff
@@ -133,8 +134,10 @@ fn create_app<'a, 'b>(
         .arg(
             Arg::with_name("kernel")
                 .long("kernel")
-                .help("Path to loaded kernel. This may be a kernel or firmware that supports a PVH \
-                entry point, a vmlinux ELF file or a Linux bzImage or achitecture equivalent")
+                .help(
+                    "Path to loaded kernel. This may be a kernel or firmware that supports a PVH \
+                entry point, a vmlinux ELF file or a Linux bzImage or achitecture equivalent",
+                )
```
Comment (sameo) [inline src/main.rs:140]: nit: That's an unrelated formatting change

**Review by sameo — APPROVED** [review]

File: src/main.rs (hunk 4 above)
Comment (sboeuf) [reply src/main.rs:140]: I didn't change this, I think it was `cargo fmt`.

**Review by liuw — APPROVED** [review]

File: src/main.rs (hunk 4 above)
Comment (sboeuf) [reply src/main.rs:140]: That's weird because `cargo fmt` does not change anything on `master` but it does with my PR. So I think this is needed, otherwise the CI is going to complain that the code is not properly formatted.
